execution/tvf/percentile_watermark_generator.go

```diff
@@ -0,0 +1,148 @@
+package tvf
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/cube2222/octosql"
+	"github.com/cube2222/octosql/execution"
+	"github.com/cube2222/octosql/streaming/storage"
+)
+
+type PercentileWatermarkGenerator struct {
+	source     execution.Node
+	timeField  octosql.VariableName
+	events     execution.Expression
+	percentile execution.Expression
+}
+
+func NewPercentileWatermarkGenerator(source execution.Node, timeField octosql.VariableName, events, percentile execution.Expression) *PercentileWatermarkGenerator {
+	return &PercentileWatermarkGenerator{
+		source:     source,
+		timeField:  timeField,
+		events:     events,
+		percentile: percentile,
+	}
+}
+
+func (w *PercentileWatermarkGenerator) Get(ctx context.Context, variables octosql.Variables, streamID *execution.StreamID) (execution.RecordStream, *execution.ExecOutput, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	sourceStreamID, err := execution.GetSourceStreamID(tx.WithPrefix(streamID.AsPrefix()), octosql.MakePhantom())
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source stream ID")
+	}
+
+	source, _, err := w.source.Get(ctx, variables, sourceStreamID) // we don't need execOutput here since we become new watermark source
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get source")
+	}
+
+	events, err := w.events.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark events")
+	}
+	if events.GetType() != octosql.TypeDuration || events.AsInt() < 1 {
+		return nil, nil, errors.Errorf("invalid watermark events: %v", events)
+	}
+
+	percentile, err := w.percentile.ExpressionValue(ctx, variables)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "couldn't get watermark percentile")
+	}
+	if percentile.GetType() != octosql.TypeDuration || percentile.AsInt() < 1 || percentile.AsInt() > 99 {
+		return nil, nil, errors.Errorf("invalid watermark percentile: %v", percentile)
+	}
+
+	ws := &PercentileWatermarkGeneratorStream{
+		source:     source,
+		timeField:  w.timeField,
+		events:     events.AsInt(),
+		percentile: percentile.AsInt(),
+		eventsSeen: 0,
+	}
+
+	return ws, execution.NewExecOutput(ws), nil // watermark generator stream now indicates new watermark source
+}
+
+type PercentileWatermarkGeneratorStream struct {
```

Comment:
Could you add a short comment that explains how this generator works? I understand that the `events` field is user given and is the amount of records that one suspects (?) and the `percentile` is after what percentile a watermark should be sent, but I am not sure I understand it 100%